### ISSUE TITLE
Metrics RBAC in wrong namespace when ServiceMonitor is launched in non-release namespace using helm

### ISSUE THREAD

**Issue by rct44** [clastix/capsule#696]
# Bug description

When deploying both capsule and capsule proxy via the provided helm charts with service monitor enabled, if the service monitor is launched into a namespace other than the capsule release namespace, the role and role binding that are created to give a service account permissions to the services/pods/endpoints are created in the service monitor namespace rather than the capsule release namespace. This means that the prometheus service account does not have the required permissions to access services/pods/endpoints in the capsule release namespace. 

See https://github.com/clastix/capsule/blob/master/charts/capsule/templates/metrics-rbac.yaml

# How to reproduce

1. Have the prometheus operator installed on your cluster, and a prometheus instance running
2. Create a values.yaml for the capsule/capsule-proxy helm chart
3. Set serviceMonitor.enabled to true
4. Set serviceMonitor.namespace to a value other than the capsule release namespace
5. Set serviceMonitor.serviceAccount.name and serviceMonitor.serviceAccount.namespace to point to the appropriate SA for your prometheus instance
6. Deploy capsule and/or capsule proxy helm chart to cluster

# Expected behavior

The helm chart automatically creates a role and role binding that gives the prometheus service account permissions to get/list/watch pods, services and endpoints in the capsule release namespace

# Logs
Note the name of the prometheus service account has been redacted and replaced with <prometheus service account>
```
ts=2023-02-03T01:19:26.934Z caller=log.go:168 level=error component=k8s_client_runtime func=ErrorDepth msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:448: Failed to watch *v1.Service: failed to list *v1.Service: services is forbidden: User \"<prometheus service account>\" cannot list resource \"services\" in API group \"\" in the namespace \"capsule-system\""
ts=2023-02-03T01:19:36.238Z caller=log.go:168 level=error component=k8s_client_runtime func=ErrorDepth msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:447: Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: endpoints is forbidden: User \"<prometheus service account>\" cannot list resource \"endpoints\" in API group \"\" in the namespace \"capsule-system\""
ts=2023-02-03T01:19:36.307Z caller=log.go:168 level=error component=k8s_client_runtime func=ErrorDepth msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:449: Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User \"<prometheus service account>\" cannot list resource \"pods\" in API group \"\" in the namespace \"capsule-system\""

```

# Additional context

- Capsule-Proxy version: 0.3.2
- Helm Chart version: 0.3.3
- OpenShift version: 4.10

**Comment by prometherion** [comment]
This should be easy to fix, if I'm correct, we just need to specify the Namespace of these resources matching the ServiceMonitor one, do you agree?

**Comment by prometherion** [comment]
But we're actually doing that: 

https://github.com/clastix/capsule/blob/d92f1e7825d3fc09f9492bf494057d22cafc71c1/charts/capsule/templates/metrics-rbac.yaml

@ptx96 @MaxFedotov may I ask your help on this to figure out why is needed?

**Comment by MaxFedotov** [comment]
hmm, seems like it was added with this commit https://github.com/clastix/capsule/commit/fc8f5a08fe49ce1884bce8049354d93849a6d63a by @ptx96 

We are not using it, because we have a clusterrole for prometheus operator with all necessary permissions

**Comment by rct44** [comment]
@prometherion - the namespace needs to match the release, not the service monitor. I think it just needs to be changed so that the namespace value is .Release.Namespace always, rather than currently where it gets overridden if the serviceMonitor.namespace value is set. It will need to change in the role and rolebinding, in both the capsule and capsule-proxy metrics-rbac.yaml files.

**Comment by prometherion** [comment]
It seems your Prometheus is running without the required RBAC to scrape metrics from the OCP/Kubernetes cluster by querying for Service, Pods, and Endpoints/Endpointslice objects.

I just wanted to give try to reproduce your issue using a vanilla Kubernetes cluster, using the Helm chart provided by Bitnami:

```
$: helm upgrade --install kube-prometheus bitnami/kube-prometheus --namespace monitoring-system --create-namespace

$: kubectl get clusterrole kube-prometheus-kube-state-metrics-monitoring-system -o yaml | grep -i endpoints
- apiGroups:
  - ""
  resources:
  - endpoints
  verbs:
  - list
  - watch

$: kubectl get clusterrole kube-prometheus-kube-state-metrics-monitoring-system -o yaml | grep -i pods
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - list
  - watch
```

The basic installation is already fulfilled the required RBAC, and if I try to install Capsule from the latest Chart and delete the metrics RBAC, it works as expected.

```
$: helm upgrade --install capsule clastix/capsule --namespace capsule-system --create-namespace --set serviceMonitor.enabled=true

$: kubectl delete clusterrole capsule-metrics-reader
```

![image](https://user-images.githubusercontent.com/9697646/217230902-4a6dce15-f183-4ab7-9380-fc3695794021.png)

tl;dr; I think the issue is not related to Capsule, but rather, to the wrong RBAC of your Prometheus instance, going to use this issue to remove the unnecessary RBAC for the metrics since those are useless, and breaking the responsibility principle trying to solve a wrong installation of a third party component.

**Comment by rct44** [comment]
Yes that works. Our prometheus instance is set up with individual namespace roles rather than cluster level permissions. We didn't initially create a separate role in this case because one was already present in the helm chart, but creating a separate one is what we usually do and was how we got around the issue once we found it. It just leaves us with the additional resources from the chart that aren't doing anything. So removing the role from the chart works just as well as a solution, so that it is at least clear that the prometheus permissions need to be set separately.
